pkg/fixer/fixer.go

```diff
@@ -0,0 +1,236 @@
+)
+
+// Report contains updated file contents and summary information about the fixes that were applied
+// during a fix operation.
+type Report struct {
+	totalFixes          int
+	fileFixedViolations map[string]map[string]struct{}
+	fileContents        map[string][]byte
+}
+
+func NewReport() *Report {
+	return &Report{
+		fileFixedViolations: make(map[string]map[string]struct{}),
+		fileContents:        make(map[string][]byte),
+	}
+}
+
+func (r *Report) SetFileContents(file string, content []byte) {
+	r.fileContents[file] = content
+}
+
+func (r *Report) GetFileContents(file string) ([]byte, bool) {
+	content, ok := r.fileContents[file]
+
+	return content, ok
+}
+
+func (r *Report) SetFileFixedViolation(file string, violation string) {
+	if _, ok := r.fileFixedViolations[file]; !ok {
+		r.fileFixedViolations[file] = make(map[string]struct{})
+	}
+
+	_, ok := r.fileFixedViolations[file][violation]
+	if !ok {
+		r.fileFixedViolations[file][violation] = struct{}{}
+		r.totalFixes++
+	}
+}
+
+func (r *Report) FileContents() map[string][]byte {
+	return r.fileContents
+}
+
+func (r *Report) FixedFiles() []string {
+	fixedFiles := make([]string, 0)
+	for file := range r.fileContents {
+		fixedFiles = append(fixedFiles, file)
+	}
+
+	// sort the files for deterministic output
+	slices.Sort(fixedFiles)
+
+	return fixedFiles
+}
+
+func (r *Report) FixedViolationsForFile(file string) []string {
+	fixedViolations := make([]string, 0)
+	for violation := range r.fileFixedViolations[file] {
+		fixedViolations = append(fixedViolations, violation)
+	}
+
+	// sort the violations for deterministic output
+	slices.Sort(fixedViolations)
+
+	return fixedViolations
+}
+
+func (r *Report) TotalFixes() int {
+	// totalFixes is incremented for each unique violation that is fixed
+	return r.totalFixes
+}
+
+// NewDefaultFixes returns a list of default fixes that are applied by the fix command.
+// When a new fix is added, it should be added to this list.
+func NewDefaultFixes() []fixes.Fix {
+	return []fixes.Fix{
+		&fixes.Fmt{},
+		&fixes.Fmt{
+			KeyOverride: "use-rego-v1",
+			OPAFmtOpts: format.Opts{
+				RegoVersion: ast.RegoV0CompatV1,
+			},
+		},
+		&fixes.UseAssignmentOperator{},
+		&fixes.NoWhitespaceComment{},
+	}
+}
+
+type Fixer struct {
+	registeredFixes map[string]any
+}
+
+func (f *Fixer) RegisterFixes(fixes ...fixes.Fix) {
+	if f.registeredFixes == nil {
+		f.registeredFixes = make(map[string]any)
+	}
+
+	for _, fix := range fixes {
+		f.registeredFixes[fix.Key()] = fix
+	}
+}
+
+func (f *Fixer) GetFixForKey(key string) (fixes.Fix, bool) {
+	fix, ok := f.registeredFixes[key]
+	if !ok {
+		return nil, false
+	}
+
+	fixInstance, ok := fix.(fixes.Fix)
+	if !ok {
+		return nil, false
+	}
+
+	return fixInstance, true
+}
+
+// OrderedFixes returns the fixes in the order they should be applied.
+// Fixes that are marked as WholeFile are applied last since they
+// can add new lines that would affect the fixes for line based violations.
```

Comment:
Hmm, I see your point. But also — if run last, it's possible that the formatter itself introduces some changes that makes it non-compliant with the fixes that got applied before it. As an example:

```rego
allow {
	true
}
```
will be rewritten to be:
```rego
allow = true
```
I don't think there are many examples of this, but it's obviously a problem if running `regal fix` still leaves some files with violations 🤔

Reply:
Ahh I hadn't used that as an example, where only opa-fmt is needed and not rego.v1.

```
package wow

import rego.v1

allow {
    true
}
```

Will leave violations now...

I wonder if for the fixing, we could only run the RegoV0CompatV1 formatter?

Reply:
Changes from today impl this now. Namely: https://github.com/StyraInc/regal/pull/653/commits/5f42a5ceb9a3f5208c98995bc094942b6bf71cb8